### PR TITLE
CPM-557: Add Parent Intents

### PR DESCRIPTION
**Pull request by Balamung**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

**Review by mmetayer — COMMENTED** [review]
no validation?

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        $this->clearDoctrineUoW();
+
+        $product = $this->productRepository->findOneByIdentifier('variant_product');
+        Assert::assertNotNull($product);
+        Assert::assertSame('variant_product', $product->getIdentifier());
+        Assert::assertEqualsCanonicalizing(null, $product->getParent());
+    }
+
+    /** @test */
+    public function it_changes_parent_to_another_family_by_clearing_parent_first():void
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php:89]:
```suggestion
    public function it_changes_parent_to_another_family_by_clearing_parent_first(): void
```

File: src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/RemoveParent.php [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Akeneo\Pim\Enrichment\Product\API\Command\UserIntent;
+
+use Webmozart\Assert\Assert;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class RemoveParent implements ParentUserIntent
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/RemoveParent.php:13]: Convert to simple product

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/RemoveParentApplier.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+use Akeneo\Pim\Enrichment\Product\API\Command\UserIntent\RemoveParent;
+use Akeneo\Pim\Enrichment\Product\API\Command\UserIntent\UserIntent;
+use Akeneo\Tool\Component\StorageUtils\Updater\ObjectUpdaterInterface;
+use Webmozart\Assert\Assert;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class RemoveParentApplier implements UserIntentApplier
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/RemoveParentApplier.php:18]: ConvertToSimpleProductApplier

File: src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/SetParent.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace Akeneo\Pim\Enrichment\Product\API\Command\UserIntent;
+
+use Webmozart\Assert\Assert;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class SetParent implements ParentUserIntent
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/SetParent.php:13]: ChangeParent

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/SetParentApplier.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+{
+    public function __construct(
+        private ObjectUpdaterInterface $productUpdater,
+    ) {
+    }
+
+    public function apply(UserIntent $userIntent, ProductInterface $product, int $userId): void
+    {
+        Assert::isInstanceOf($userIntent, SetParent::class);
+
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/SetParentApplier.php:27]: Early return if parent is already this product model

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    /** @test */
+    public function it_can_add_change_and_remove_a_parent(): void
+    {
+        $this->createProductModel('root', 'color_variant_accessories', [
+            'categories' => ['print'],
+        ]);
+        $this->createProductModel('root2', 'color_variant_accessories', [
+            'categories' => ['print'],
+        ]);
+        $this->createProduct('variant_product', [
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php:41]: use Service API

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+            userId: $this->getUserId('betty'),
+            productIdentifier: 'variant_product',
+            parentUserIntent: new RemoveParent()
+        );
+        $this->messageBus->dispatch($command);
+        $this->clearDoctrineUoW();
+
+        $product = $this->productRepository->findOneByIdentifier('variant_product');
+        Assert::assertNotNull($product);
+        Assert::assertSame('variant_product', $product->getIdentifier());
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php:84]: not needed?

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        Assert::assertEqualsCanonicalizing(null, $product->getParent());
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php:85]: Assert::assertFalse(productisvariant)

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        ]);
+
+        $this->createProductModel('root', 'color_variant_accessories', [
+            'categories' => ['print'],
+        ]);
+        $this->createProductModel('root2', 'size_variant_clothes', [
+            'categories' => ['print'],
+        ]);
+
+        $this->createProduct('variant_product', [
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php:112]: serviceAPI

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+            productIdentifier: 'variant_product',
+            parentUserIntent: new SetParent('unknown')
+        );
+
+        $this->expectException(ViolationsException::class);
+        $this->expectExceptionMessage('Property "parent" expects a valid parent code. The parent product model does not exist, "unknown" given.');
+
+        $this->messageBus->dispatch($command);
+    }
+
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php:176]: Add test : remove parent on simple product

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/API/Command/UserIntent/SetParentSpec.php [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class SetParentSpec extends ObjectBehavior
+{
+    function it_can_be_constructed_with_parent_code()
+    {
+        $this->beConstructedWith('test_product_model');
+        $this->shouldBeAnInstanceOf(SetParent::class);
+        $this->parentCode()->shouldReturn('test_product_model');
+    }
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/API/Command/UserIntent/SetParentSpec.php:21]: test assert

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Application/Applier/SetParentApplierSpec.php [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+    function it_applies_set_parent_user_intent(ObjectUpdaterInterface $updater): void
+    {
+        $product = new Product();
+        $setParent = new SetParent('product_model_code');
+        $updater->update($product,['parent' => 'product_model_code'])->shouldBeCalledOnce();
+
+        $this->apply($setParent, $product, 1);
+    }
+
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Application/Applier/SetParentApplierSpec.php:34]: Test if same parent

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php (hunk 10 above)
Comment (Balamung) [reply src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php:176]: I didn't add the test in the end, as the other product creations through the serviceAPI call this intent on creation to ensure products are simple products and not variants

File: src/Akeneo/Pim/Enrichment/Product/back/Application/UpsertProductHandler.php [hunk 13]
```diff
@@ -86,6 +86,7 @@ private function updateProduct(ProductInterface $product, UpsertProductCommand $
                     'enabledUserIntent' => $command->enabledUserIntent(),
                     'familyUserIntent' => $command->familyUserIntent(),
                     'categoryUserIntent' => $command->categoryUserIntent(),
+                    'parentUserIntent' => $command->parentUserIntent(),
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/UpsertProductHandler.php:89]: to check: I believe the parentUserIntent should be handled at the very beginning. Can you create a card to check that, please?

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+        $this->messageBus->dispatch($command);
+        $this->clearDoctrineUoW();
+
+        $product = $this->productRepository->findOneByIdentifier('variant_product');
+        Assert::assertNotNull($product);
+        Assert::assertFalse($product->isVariant());
+    }
+
+    /** @test */
+    public function it_changes_parent_to_another_family_by_clearing_parent_first(): void
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php:95]:
```suggestion
    public function it_cannot_change_a_parent_from_a_different_family_variant(): void
```

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Application/Applier/ConvertToSimpleProductApplierSpec.php [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    ): void
+    {
+        $removeParentIntent = new ConvertToSimpleProduct();
+        $product->getParent()->willReturn($productModel);
+        $removeParent->from($product)->shouldBeCalledOnce();
+
+        $this->apply($removeParentIntent, $product, 1);
+    }
+
+    function it_throws_an_exception_when_product_has_no_parent(
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Application/Applier/ConvertToSimpleProductApplierSpec.php:41]: it does not throw an exception here ??

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/ConvertToSimpleProductApplier.php [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+{
+    public function __construct(
+        private RemoveParentInterface $removeParent
+    ) {
+    }
+
+    public function apply(UserIntent $userIntent, ProductInterface $product, int $userId): void
+    {
+        Assert::isInstanceOf($userIntent, ConvertToSimpleProduct::class);
+        if (null === $product->getParent()) {
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/ConvertToSimpleProductApplier.php:28]:
```suggestion
        if (!$product->isVariant()) {
```
(it means exactly the same, but I find it's more understandable that way)

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php (hunk 10 above)
Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php:193]: I don't understand :p, but really this test is not necessary as we don't apply the ConvertToSimpleProduct for non variant products, so the spec on the applier is enough

**Review by mmetayer — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php (hunk 14 above)
Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductVariantIntegration.php:95]: I would suggest:

```suggestion
    public function it_can_only_change_parent_to_another_family_by_clearing_parent_first(): void
```

**Review by nmarniesse — APPROVED** [review]

**Pull request merged**